File: tui/internal/ui/executor.go
package ui

import (
	"fmt"

	"hejje.money/tui/internal/api"
)

// RenderExecutor is one line: this instance's role and who is active.
func RenderExecutor(s api.ExecutorStatus) string {
	switch s.Role {
	case "NOT_REQUIRED":
		return fmt.Sprintf("single instance %s (lease not required)\n", s.Instance)
	case "ACTIVE":
		return fmt.Sprintf("ACTIVE %s (epoch %d)\n", s.Instance, s.Epoch)
	}
	active := "none (lease expired)"
	if s.ActiveInstance != "" && s.ActiveEpoch != nil {
		active = fmt.Sprintf("%s (epoch %d)", s.ActiveInstance, *s.ActiveEpoch)
	}
	line := fmt.Sprintf("STANDBY %s — active: %s", s.Instance, active)
	if s.HoldUntil != nil {
		line += "; holding back until " + s.HoldUntil.Local().Format("15:04:05")
	}
	return line + "\n"
}

File: web/src/lib/executor.ts
export interface ExecutorStatus {
  instance: string; role: 'ACTIVE' | 'STANDBY' | 'NOT_REQUIRED'; required: boolean; held: boolean; epoch: number;
  activeInstance?: string | null; activeEpoch?: number | null; expiresAt?: string | null; holdUntil?: string | null;
}
export interface BrokerAccount { id: string; broker: string; accountId: string; label?: string | null; active: boolean; activatedAt?: string | null; activatedBy?: string | null; }
export interface BrokerAccountsView { adapter: string; active: BrokerAccount | null; accounts: BrokerAccount[]; }

/** "ACTIVE (epoch 3)" / "STANDBY — active: vm-b (epoch 4)" / "single instance (lease not required)". */
export function executorLine(s: ExecutorStatus): string {
  if (s.role === 'NOT_REQUIRED') return `single instance ${s.instance} (lease not required)`;
  if (s.role === 'ACTIVE') return `ACTIVE ${s.instance} (epoch ${s.epoch})`;
  const active = s.activeInstance ? `${s.activeInstance} (epoch ${s.activeEpoch})` : 'none (lease expired)';
  return `STANDBY ${s.instance} — active: ${active}${s.holdUntil ? `; holding back until ${new Date(s.holdUntil).toLocaleTimeString()}` : ''}`;
}

/** Failover is offered only on the instance that holds the lease. */
export function canFailover(s?: ExecutorStatus): boolean {
  return !!s && s.required && s.role === 'ACTIVE';
}
